crates/ruff_linter/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs

```diff
@@ -57,46 +67,85 @@ pub(crate) fn unnecessary_coding_comment(
 ) {
     // The coding comment must be on one of the first two lines. Since each comment spans at least
     // one line, we only need to check the first two comments at most.
-    for comment_range in comment_ranges.iter().take(2) {
-        // If leading content is not whitespace then it's not a valid coding comment e.g.
-        // ```
-        // print(x) # coding=utf8
-        // ```
-        let line_range = locator.full_line_range(comment_range.start());
-        if !locator
-            .slice(TextRange::new(line_range.start(), comment_range.start()))
-            .trim()
-            .is_empty()
-        {
-            continue;
-        }
+    let coding_comments = comment_ranges
+        .iter()
+        .take(2)
+        .map(|comment_range| coding_comment(locator, indexer, *comment_range))
+        .collect::<Vec<_>>();
 
-        // If the line is after a continuation then it's not a valid coding comment e.g.
-        // ```
-        // x = 1 \
-        //    # coding=utf8
-        // x = 2
-        // ```
-        if indexer
-            .preceded_by_continuations(line_range.start(), locator.contents())
-            .is_some()
-        {
-            continue;
+    match &coding_comments[..] {
+        [Some(CodingComment::UTF8(ranges))]
+        | [Some(CodingComment::UTF8(ranges)), None]
+        | [None, Some(CodingComment::UTF8(ranges))] => {
+            report(diagnostics, ranges.line_range, ranges.self_range);
         }
 
-        if CODING_COMMENT_REGEX.is_match(locator.slice(line_range)) {
-            #[allow(deprecated)]
-            let index = locator.compute_line_index(line_range.start());
-            if index.to_zero_indexed() > 1 {
-                continue;
-            }
-
-            let mut diagnostic = Diagnostic::new(UTF8EncodingDeclaration, *comment_range);
-            diagnostic.set_fix(Fix::safe_edit(Edit::deletion(
-                line_range.start(),
-                line_range.end(),
-            )));
-            diagnostics.push(diagnostic);
+        [Some(CodingComment::UTF8(ranges_1)), Some(CodingComment::UTF8(ranges_2))] => {
+            report(diagnostics, ranges_1.line_range, ranges_1.self_range);
+            report(diagnostics, ranges_2.line_range, ranges_2.self_range);
```

Comment:
If this block removes both UTF-8 encoding declarations, what happens if there is a third, non-UTF-8 encoding declaration?
```python
# coding=utf-8
# coding=utf-8
# coding=ascii
```

I think it would be better to only offer the fix for `ranges_1`. If it safe to delete `ranges_2` too, that can happen on the next iteration.

Reply:
In that case, we shouldn't raise a diagnostic for the second encoding comment.